### ISSUE TITLE
Method ambiguity on 0.7

### ISSUE THREAD

**Issue by hessammehr** [willtebbutt/Stheno.jl#11]
Running the first sample code block on 0.7 gives:
```
ERROR: MethodError: *(::LinearAlgebra.Adjoint{Float64,LinearAlgebra.UpperTriangular{Float64,BlockArrays.BlockArray{Float64,2,AbstractArray{Float64,2}}}}, ::BlockArrays.BlockArray{Float64,2,Array{Float64,2}}) is ambiguous. Candidates:
  *(A::AbstractArray{T,2}, B::BlockArrays.AbstractBlockArray{T,2}) where T in Stheno at C:\Users\abcd\.julia\dev\Stheno\src\util\block_arrays.jl:280
  *(adjA::LinearAlgebra.Adjoint{#s608,#s607} where #s607<:LinearAlgebra.AbstractTriangular where #s608, B::AbstractArray{T,2} where T) in LinearAlgebra at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\LinearAlgebra\src\triangular.jl:1813
Possible fix, define
  *(::LinearAlgebra.Adjoint{T,#s607} where #s607<:LinearAlgebra.AbstractTriangular, ::BlockArrays.AbstractBlockArray{T,2})
Stacktrace:
 [1] rand(::Random.MersenneTwister, ::BlockGP{Array{GP{FiniteMean,FiniteKernel},1}}, ::Int64) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:35
 [2] rand(::Random.MersenneTwister, ::Array{GP{FiniteMean,FiniteKernel},1}, ::Int64) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:46
 [3] rand(::Random.MersenneTwister, ::Array{GP{FiniteMean,FiniteKernel},1}) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:45
```
The error is triggered by the 
I tried a hacky fix:
```julia
# block_arrays.jl, after line 280
*(A::LinAlg.Adjoint{T,M} where M<:LinAlg.AbstractTriangular, B::ABM{T}) where T = BlockMatrix([A]) * B
```

And that seemed to take care of the ambiguity but then I got the following on re-running the code sample:
```
ERROR: AssertionError: are_conformal(A, B)
Stacktrace:
 [1] *(::BlockArrays.BlockArray{Float64,2,Array{Float64,2}}, ::BlockArrays.BlockArray{Float64,2,Array{Float64,2}}) at C:\Users\abcd\.julia\dev\Stheno\src\util\block_arrays.jl:268
 [2] *(::LinearAlgebra.Adjoint{Float64,LinearAlgebra.UpperTriangular{Float64,BlockArrays.BlockArray{Float64,2,AbstractArray{Float64,2}}}}, ::BlockArrays.BlockArray{Float64,2,Array{Float64,2}}) at C:\Users\abcd\.julia\dev\Stheno\src\util\block_arrays.jl:281
 [3] rand(::Random.MersenneTwister, ::BlockGP{Array{GP{FiniteMean,FiniteKernel},1}}, ::Int64) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:35
 [4] rand(::Random.MersenneTwister, ::Array{GP{FiniteMean,FiniteKernel},1}, ::Int64) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:46
 [5] rand(::Random.MersenneTwister, ::Array{GP{FiniteMean,FiniteKernel},1}) at C:\Users\abcd\.julia\dev\Stheno\src\gp\block_gp.jl:45
 [6] top-level scope at none:0
```

It would be a good starting point to know which method specialisation we would like to go with here (`block_array.jl` or `triangular.jl`).

**Comment by willtebbutt** [comment]
Thanks for giving this a go. I've not started to upgrade for 0.7 yet (time constraints), so I would be very surprised if it is possible to get even a basic example working under the current code base.

Quickly looking at this error, I'm not sure that the above fix is going to be appropriate. I'll try and figure out what is going on here some time this week, but I'm quite pressed for time currently so can't make any promises :(

I would recommend running on 0.6 for the time being.

**Comment by hessammehr** [comment]
My understanding is that the ambiguity from an undecidedable situation like the following

```julia
method(::A, ::BB) where BB<:B
method(::AA, ::B) where AA<:A
```

Adding a third method that is more specialised than both of the above will resolve the ambiguity.

```julia
method(::AA, ::BB) where {BB<:B,AA<:A}
```

That said, I'm not sure why applying this fix goes on to break the `are_conformal` check later on.

**Comment by willtebbutt** [comment]
Apologies, I should have been clearer. That's certainly the cause of the error, I'm just not completely sure what the correct implementation is for the method that resolves the ambiguity. The source of the error in `are_conformal` is to do with block sizes of block matrices not matching up in the correct way, which is probably a result of the fix you've proposed. (Something like) the following hack might work:
```
function *(A::Adjoint{T,<:AbstractTriangular}, B::ABM{T}) where T
    return Ac_mul_B(A.parent, B)
end
```
which should take advantage of the existing `Ac_mul_B` method that I implemented for 0.6 for block matrix multiplication.

Again, this kind of issue is likely to crop up a lot, so this probably isn't an isolated issue.

**Comment by hessammehr** [comment]
So I tried your suggestion and it kind of worked - almost. Now I get an error for the following line:

```julia
f₁′Xp, f₂′Xp, f₃′Xp = rand(rng, [f₁′(Xp), f₂′(Xp), f₃′(Xp)], 25)
```

 I realize that you are busy but for the record or for anyone who might be interested, here is the error that I get now:

```
ERROR: MethodError: no method matching AbstractArray{T,2} where T(::LazyPDMat{Float64,ToeplitzMatrices.SymmetricToeplitz{Float64}})
Closest candidates are:
  AbstractArray{T,2} where T(::Any) where T<:AbstractArray at abstractarray.jl:22
  AbstractArray{T,2} where T(::Union{LinearAlgebra.QR, LinearAlgebra.QRCompactWY}) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\LinearAlgebra\src\qr.jl:382
  AbstractArray{T,2} where T(::LinearAlgebra.QRPivoted) at C:\cygwin\home\Administrator\buildbot\worker\package_win64\build\usr\share\julia\stdlib\v0.7\LinearAlgebra\src\qr.jl:389
  ...
Stacktrace:
 [1] AbstractArray{T,2} where T(::LazyPDMat{Float64,ToeplitzMatrices.SymmetricToeplitz{Float64}}) at .\abstractarray.jl:22
 [2] _pairwise(::Stheno.ConditionalKernel, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\conditional.jl:71
 [3] pairwise(::Stheno.ConditionalKernel, ::StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}}) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\kernel.jl:90
 [4] _pairwise(::FiniteKernel, ::Base.OneTo{Int64}) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\finite.jl:52
 [5] pairwise(::FiniteKernel, ::Base.OneTo{Int64}) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\kernel.jl:50
 [6] pairwise(::BlockKernel, ::BlockData{Int64,Base.OneTo{Int64}}) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\block.jl:83
 [7] AbstractArray{T,2} where T(::BlockKernel) at C:\Users\Uigedail\.julia\dev\Stheno\src\mean_and_kernel\conversion.jl:19
 [8] cov at C:\Users\Uigedail\.julia\dev\Stheno\src\gp\abstract_gp.jl:50 [inlined]
 [9] rand(::Random.MersenneTwister, ::BlockGP{Array{GP{FiniteMean,FiniteKernel},1}}, ::Int64) at C:\Users\Uigedail\.julia\dev\Stheno\src\gp\block_gp.jl:35
 [10] rand(::Random.MersenneTwister, ::Array{GP{FiniteMean,FiniteKernel},1}, ::Int64) at C:\Users\Uigedail\.julia\dev\Stheno\src\gp\block_gp.jl:46
```

**Comment by willtebbutt** [comment]
Closed by merging #12
